templates/15_insert.go.tpl

```diff
@@ -201,3 +201,173 @@ CacheNoHooks:
+		{{- if eq $col.Name "updated_at" -}}
+		{{- if eq $col.Type "time.Time"}}
+		if row.UpdatedAt.IsZero() {
+			row.UpdatedAt = currTime
+		}
+		{{- else}}
+		if queries.MustTime(row.UpdatedAt).IsZero() {
+			queries.SetScanner(&row.UpdatedAt, currTime)
+		}
+		{{- end -}}
+		{{- end -}}
+		{{end}}
+		{{end}}
+		{{- end}}
+ 		nzDefaults := queries.NonZeroDefaultSet({{$alias.DownSingular}}ColumnsWithDefault, row)
+ 		wl, returnColumns := columns.InsertColumnSet(
+			{{$alias.DownSingular}}Columns,
+			{{$alias.DownSingular}}ColumnsWithDefault,
+			{{$alias.DownSingular}}ColumnsWithoutDefault,
+			nzDefaults,
+		)
+ 		if i == 0 {
+			sql = "INSERT INTO {{$schemaTable}} " + "({{.LQ}}" + strings.Join(wl, "{{.RQ}},{{.LQ}}") + "{{.RQ}})" + " VALUES "
+			{{if not .Dialect.UseLastInsertID -}}
+			retQuery = " RETURNING {{.LQ}}" + strings.Join(returnColumns, "{{.RQ}},{{.LQ}}") + "{{.RQ}} "
+			{{else -}}
+			retQuery = fmt.Sprintf("SELECT {{.LQ}}%s{{.RQ}} FROM {{$schemaTable}} WHERE {{.LQ}}%s{{.RQ}} IN %s",
+			  strings.Join(returnColumns, "{{.RQ}},{{.LQ}}"),
+			  {{$alias.DownSingular}}PrimaryKeyColumns[0],
+			  strmangle.Placeholders(dialect.UseIndexPlaceholders, len(o), 1, len(o)))
+			{{end -}}
+		}
+ 		sql += strmangle.Placeholders(dialect.UseIndexPlaceholders, len(wl), len(vals)+1, len(wl))
+		if i != len(o)-1 {
+			sql += ","
+		}
+ 		valMapping, err := queries.BindMapping({{$alias.DownSingular}}Type, {{$alias.DownSingular}}Mapping, wl)
+		if err != nil {
+			return err
+		}
+		retMapping, err := queries.BindMapping({{$alias.DownSingular}}Type, {{$alias.DownSingular}}Mapping, returnColumns)
+		if err != nil {
+			return err
+		}
+		value := reflect.Indirect(reflect.ValueOf(row))
+		vals = append(vals, queries.ValuesFromMapping(value, valMapping)...)
+		retMaps = append(retMaps, queries.PtrsFromMapping(value, retMapping))
+	}
+ 	{{if not .Dialect.UseLastInsertID -}}
+	sql += retQuery
+	{{end -}}
+ 	if boil.DebugMode {
+		fmt.Fprintln(boil.DebugWriter, sql)
+		fmt.Fprintln(boil.DebugWriter, vals...)
+	}
+ 	{{if .Dialect.UseLastInsertID -}}
+	{{- $canLastInsertID := .Table.CanLastInsertID -}}
+	{{if $canLastInsertID -}}
+		{{if .NoContext -}}
+	result, err := exec.Exec(ctx, sql, vals...)
+		{{else -}}
+	result, err := exec.ExecContext(ctx, sql, vals...)
+		{{end -}}
+	{{else -}}
+		{{if .NoContext -}}
+	_, err := exec.Exec(ctx, sql, vals...)
+		{{else -}}
+	_, err := exec.ExecContext(ctx, sql, vals...)
+		{{end -}}
+	{{- end}}
+	if err != nil {
+		return errors.Wrap(err, "{{.PkgName}}: unable to insert into {{.Table.Name}}")
+	}
+	{{if $canLastInsertID -}}
+	var lastID int64
+	{{- end}}
+ 	{{if $canLastInsertID -}}
+	lastID, err = result.LastInsertId()
+  	if err != nil {
+		return ErrSyncFail
+	}
+ 	{{$colName := index .Table.PKey.Columns 0 -}}
+	{{- $col := .Table.GetColumn $colName -}}
+	{{- $colTitled := $colName | titleCase -}}
+ 	ids := []interface{}{}
+	for i := 0; i < len(o); i++ {
+		id := {{$col.Type}}(lastID)+{{$col.Type}}(i)
```

Comment:
I haven't contributed before, so my apologies if this is off-base.

In MySQL, there is no general way to correctly fetch generated auto incrementing IDs when you insert more than one row in the same statement. There are ways to do it if some MySQL options have the right settings: you need to know the value of `auto_increment_increment` (so you can skip numbers accordingly), and you need to have `innodb_autoinc_lock_mode` set to 1 or 0 (so the IDs being generated are consecutive; note that the default in MySQL 8 is non-consecutive, for performance reasons). You could also write lock the table to guarantee you're the only one generating IDs, but that's certainly not what most people would expect to happen.

I (and probably other people too) have to do bulk inserts for performance reasons. In AWS (and probably most cloud providers), the minimum latency for executing a single query is ~1 ms. So if you insert 10,000 rows using individual insert statements, it will take at least 10 seconds, and it will monopolize DB resources. If you do it as a bulk insert, it will take only a handful of milliseconds.

Because bulk insert support is very useful (and I suspect most people will just implement it themselves anyway if it isn't provided by the library), I suggest implementing it with the caveat that, in MySQL, you won't be able to get auto generated IDs back if any were generated.

Reply:
Well - this is now made easier now that each database driver can create extra functions that differ from the core completely. Perhaps we could use that for bulk insert? Though I'm still on the fence about it in general. It's just such a departure from what makes sqlboiler useful. If you're not doing the "get the id back" part, and binding other default values back into the struct, why use it at all for insertion?